### PR TITLE
Better error handling during force shutdown

### PR DESCRIPTION
**Pull request by wjordan**
### Description

Refactors tests covering shutdown behavior to be less flaky (the current tests depend on fragile `sleep` timing that isn't 100% deterministic), and also fixes a couple of subtle timing edge-cases around force-shutdown behavior:

- Only allow `ForceShutdown` to be raised in a thread during specific areas of the connection-processing cycle (marked by `with_force_shutdown` blocks), to ensure that the raised error is always rescued and handled cleanly. (This fixes an issue where the `force_shutdown_after: 0` option throws uncaught exceptions from the threadpool on shutdown)
- ~Add `@queue_mutex` and helper methods `if_queue_requests` and `queue_client` to `Server` to handle an edge case where the reactor shuts down while a thread is concurrently queuing a client. (#2337)~ A fix for this issue was merged in #2377.

### Your checklist for this pull request

- [x] I have reviewed the [guidelines for contributing](../blob/master/CONTRIBUTING.md) to this repository.
- [x] I have added an entry to [History.md](../blob/master/History.md) if this PR fixes a bug or adds a feature. If it doesn't need an entry to HISTORY.md, I have added `[changelog skip]` the pull request title.
- [x] I have added appropriate tests if this PR fixes a bug or adds a feature.
- [x] My pull request is 100 lines added/removed or less so that it can be easily reviewed.
- [x] If this PR doesn't need tests (docs change), I added `[ci skip]` to the title of the PR.
- [x] If this closes any issues, I have added "Closes `#issue`" to the PR description or my commit messages.
- [x] I have updated the documentation accordingly.
- [ ] All new and existing tests passed, including Rubocop.

### CODE REVIEW

File: lib/puma/client.rb [hunk 1]
```diff
@@ -244,19 +244,17 @@ def eagerly_finish
     end # IS_JRUBY
 
     def finish(timeout)
-      return true if @ready
```
Comment (nateberkopec) [inline lib/puma/client.rb:247]: Hm, what's the impact of changing the return value here? Any?

Comment (nateberkopec) [reply lib/puma/client.rb:247]: Just looking through the rest of the code, seems like we never checked the value.

File: lib/puma/server.rb [hunk 2]
```diff
@@ -344,12 +343,33 @@ def handle_check
       return false
     end
 
+    # Runs the provided block if the Reactor queue is enabled.
+    # Returns block return value if queue is enabled, or nil if queue is not enabled.
+    def if_queue_requests
+      @queue_mutex.synchronize do
+        if @queue_requests
```
Comment (nateberkopec) [inline lib/puma/server.rb:350]: Since queue_requests is set at configure-time, I think we can move this outside the mutex, right?

File: lib/puma/server.rb [hunk 3]
```diff
@@ -344,12 +343,33 @@ def handle_check
...
+        end
+      end
+    end
+    private :if_queue_requests
+
+    # Queues the provided client
+    def queue_client(client, timeout)
+      if_queue_requests do
+        client.set_timeout timeout
+        @reactor.add client
```
Comment (nateberkopec) [inline lib/puma/server.rb:361]: Would it be possible to keep concurrency as a concern of the Reactor rather than the Server? It sort of feels like we could refactor this a little further to keep concurrency concerns out of this class.

**Comment by nateberkopec** [comment]
I think my only concerns are the concurrency-awareness being added to Server when I think it hasn't had many concurrency-related reponsibilities yet. I'm wondering if we can shove all of that into Reactor instead so we keep Puma's main "concurrency-aware" classes as Reactor and Threadpool.

**Comment by nateberkopec** [comment]
Sounds good, we can block this PR on that one.

**Comment by wjordan** [comment]
Rebased after #2377 was merged, which fixed the second issue addressed by this PR (I've crossed out that part in the PR description above). This also means that the remaining part of this PR is no longer blocked on #2279 so it's ready for another review.
